src/auth/firebase.service.ts

```diff
@@ -0,0 +1,156 @@
+import * as fs from 'fs';
+import { Injectable, Inject } from '@nestjs/common';
+import { initializeApp, App, cert } from 'firebase-admin/app';
+import { getAuth } from 'firebase-admin/auth';
+import { Firestore, getFirestore, Timestamp } from 'firebase-admin/firestore';
+
+import { FirebaseOptions } from './interfaces';
+import { MODULE_OPTIONS_TOKEN } from './firebase.module-definition';
+
+@Injectable()
+export class FirebaseService {
+  app: App;
+  db: Firestore;
+
+  constructor(
+    @Inject(MODULE_OPTIONS_TOKEN)
+    options?: FirebaseOptions,
+  ) {
+    const appOptions = options
+      ? {
+          credential: cert(
+            JSON.parse(fs.readFileSync(options.credentialFilePath, 'utf8')),
+          ),
+          databaseURL: options.databaseUri,
+        }
+      : {};
+
+    this.app = initializeApp(appOptions);
+    this.db = getFirestore(this.app);
+  }
+
+  async verifyIdToken(idToken: string) {
+    return getAuth(this.app).verifyIdToken(idToken);
+  }
+
+  async getUserInfo(uid: string) {
+    return this.db.collection('users').doc(uid).get();
+  }
+
+  private upsertElementByArticleId = (
+    articleId: string,
+    list: any[],
+    element: any,
+  ) => {
+    const newArray = [...list];
+    const index = list.findIndex((s) => s.articleId === articleId);
```

Comment:
We could use list.find(…) ?newArray.push(element): …

Reply:
The index is still needed to modify the element if found. Is there a better practice to do the same thing?

Reply:
Try this:
```
const newArray = list.reduce((acc, curr) => {
    if (curr.articleId === element.articleId) {
      return [...acc, element];
    } else {
      return [...acc, curr];
    }
  }, []);

  if (!newArray.includes(element)) {
    newArray.push(element);
  }

  return newArray;
```

Reply:
Your answer is same to ChatGPT, so just choose what u want, the perfeormance are simlliar
![image](https://user-images.githubusercontent.com/21107283/227734002-f0f42ff0-825c-4176-bbf1-275807b7dd25.png)

Reply:
I choose to remain my implementation.